library/std/src/sys/uefi/os.rs

```diff
@@ -0,0 +1,357 @@
+where
+    I: Iterator<Item = T>,
+    T: AsRef<OsStr>,
+{
+    Err(JoinPathsError)
+}
+
+impl fmt::Display for JoinPathsError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        "not supported on this platform yet".fmt(f)
+    }
+}
+
+impl StdError for JoinPathsError {
+    #[allow(deprecated)]
+    fn description(&self) -> &str {
+        "not supported on this platform yet"
+    }
+}
+
+// Implemented using EFI_LOADED_IMAGE_DEVICE_PATH_PROTOCOL
+pub fn current_exe() -> io::Result<PathBuf> {
+    use r_efi::efi::protocols::{device_path, loaded_image_device_path};
+
+    match common::get_current_handle_protocol::<device_path::Protocol>(
+        loaded_image_device_path::PROTOCOL_GUID,
+    ) {
+        Some(mut x) => unsafe { super::path::device_path_to_path(x.as_mut()) },
+        None => Err(io::const_io_error!(
+            io::ErrorKind::Uncategorized,
+            "failed to acquire EFI_LOADED_IMAGE_DEVICE_PATH_PROTOCOL",
+        )),
+    }
+}
+
+// FIXME: Implement using Variable Services
+pub struct Env {
+    last_var_name: Vec<u16>,
+    last_var_guid: r_efi::efi::Guid,
+}
+
+impl Iterator for Env {
+    type Item = (OsString, OsString);
+    fn next(&mut self) -> Option<(OsString, OsString)> {
+        let mut key = self.last_var_name.clone();
+        // Only Read Shell and Rust variables. UEFI variables can be random bytes of data and so
+        // not much point in reading anything else in Env context.
+        let val = match self.last_var_guid {
+            uefi_vars::RUST_VARIABLE_GUID => {
+                uefi_vars::get_variable_utf8(&mut key, uefi_vars::RUST_VARIABLE_GUID)
+            }
+            uefi_vars::SHELL_VARIABLE_GUID => {
+                uefi_vars::get_variable_ucs2(&mut key, self.last_var_guid)
+            }
+            _ => None,
+        };
+        let (k, g) =
+            uefi_vars::get_next_variable_name(&self.last_var_name, &self.last_var_guid).ok()?;
+        self.last_var_guid = g;
+        self.last_var_name = k;
+        match val {
+            None => self.next(),
+            Some(x) => Some((OsString::from_wide(&key[..(key.len() - 1)]), x)),
+        }
+    }
+}
+
+pub fn env() -> Env {
+    // The Guid should be ignored, so just passing anything
+    let (key, guid) =
+        uefi_vars::get_next_variable_name(&[0], &uefi_vars::RUST_VARIABLE_GUID).unwrap();
+    Env { last_var_name: key, last_var_guid: guid }
+}
+
+// Tries to get variable from bot RUST_VARIABLE_GUID and SHELL_VARIABLE_GUID. Precedence:
+// RUST_VARIABLE_GUID > SHELL_VARIABLE_GUID
+pub fn getenv(key: &OsStr) -> Option<OsString> {
+    let mut k = common::to_ffi_string(key);
+    if let Some(x) = uefi_vars::get_variable_utf8(&mut k, uefi_vars::RUST_VARIABLE_GUID) {
+        Some(x)
+    } else {
+        uefi_vars::get_variable_ucs2(&mut k, uefi_vars::SHELL_VARIABLE_GUID)
+    }
+}
+
+// Only possible to set variable to RUST_VARIABLE_GUID
```

Comment:
UEFI shell supports this with `SetEnv`.

Reply:
It's possible to write to shell variables here by just replacing `RUST_VARIABLE_GUID` with `SHELL_VARIABLE_GUID`. In fact, I am already reading Shell variables when getting variables. 

However, since UEFI variables are just bytes of data instead of strings, separating Shell variables (Which need to be UCS-2) from Rust variables allows saving on allocation by storing UTF-8 directly and reading that.

Additionally, it might be helpful if we want to cleanup the variables from Rust cleanup or something.